### DISCUSSION TITLE
Use a toggle entity as power control for a player

### DISCUSSION THREAD

**Discussion by marcelveldt** [music-assistant/support#467] [Feature requests and ideas]
Add possibility to use a toggle entity (switch, light, input_boolean) as power control for a media player.
This allows for example the usecase to turn on/off an external amplifier etc. without the need to use a template media player.

**Comment by erkr** [comment]
> usecase to turn on/off an external amplifier etc

For such a use case (applicable for one of my CC's), I would like to extend the feature request to allow redirecting the volume controls (up, down, set) to the amplifier as well. Similar to the universal media player template 
Best
Eric

**Comment by marcelveldt** [comment]
Maybe we can turn it around and make MA compatible with the templated media player. I do not want to replicate features that are already in HA and instead focus on the music in this integration.
So what are we missing if we're using a "wrapped" media player from a template ? 
Should MA perhaps derive the source/control player from the template ?

I did the same with the helper groups btw, you can add a media player group to MA and all players will play the same music.

**Comment by erkr** [comment]
Hi Marcel

Good question. I already had these universal media players defined before MA. In generaal they work great to wrap another media player and have the ability to overrule particular commands (on, off, vol up down mute set...)
With MA I only needed to change the child player with the exported mass player to have media titles and background working again (only mute is now broken)

Indeed not useful to replicate that functionality. But What is the benefit to control them from MA, instead of using the mass versions in the templates?

What is missing? One issue I know of is that the universal media players don't support browsing media (error: Integration does not support browsing media). The rest seems to work properly.

Best
Eric

**Comment by marcelveldt** [comment]
And what about adding the universal media player to MA and hide them in the interface. You can then use the "output" media players generated by MA as daily driver which in term controls the (hidden) universal player. Media browsing will then also work.

Mute should be fixed soon, I added that to the lib yesterday and now I must only update a small part in the HA integration code.

original cast media_player --> hide / do not use directly
universal media_player --> wraps the cast media --> use in MA and set to be hidden in the UI
MA resulting player --> final player with all controls, metadata and media browsing 

It may sound a bit complex but I guess it is exactly what you need?
It should not bring any noticeable additional overhead because it is all eventdriven

**Comment by erkr** [comment]
I'm perfectly fine with that scenario:
pro: no extended support for media browsing is needed in the universal players 
con: I have to update my state handling in the configured universal media_players 

some explanation:  I simplified the state of these universal players to just follow the On/Off state of the power plugs: 
```
    attributes:
      state: light.kk_speakers
```
For displaying / control in the UI that is sufficient, but I guess you need all the player states, 
I commented out the above state attribute and found out that the template only follows the player states when the source CC is available:
- When switched off, the template state is off, but the CC is unavailable
- When the template is on, the state is on until the CC is on line. Than it adopts the player state instead

So the problem I foresee is how to discriminate a CC off state (is available, but not connected by a client) from the template off state (no CC available)

**Comment by marcelveldt** [comment]
I've had the same usecase as you and in the end I settled on another route and just leave the CC powered and instead control the amplifier/speaker connected to it. The CC consumes about 1.2W idle so that is not the whole world.

The more I think of it, the more I think there indeed is a usecase for this in the MA codebase as a simple alternative for the template media player approach. In fact, whole MA can be seen as an alternative for the templated media player.

So this is how I see it:

- Add a setting in MA integration to attach a toggle entity to a media player for power control.
- Add a setting in MA integration to attach another media_player for volume control.
- Any other usecases will not be supported by MA and are considered "very advanced" and must be solved by writing a template mediaplayer in yaml, which in turn can be imported by MA.

I think it's a pretty common scenario (in Home Assistant at least) that a media player is actually powered from a smart plug or something like that or in the case of a streaming device connected to an amplifier. The current way to handle that requires a fair bit of fiddling with yaml.

Other approach would be to just introduce a new helper entity in HA for this task. A simplified GUI based universal media player with the features I mentioned above: Wrap a media player and allow the power and/or volume control to be handled by another device.

To be continued...

**Comment by erkr** [comment]
Once decided, I'm happy to test it :-)
For now definitely not urgent

**Comment by erkr** [comment]
Thinking over this feature request;
If the mute service is fixed https://github.com/music-assistant/hass-music-assistant/issues/137, I basically don't need this extension. 
For browsing, I could add a shortcut to the MA UI in the Mini Media Player card

**Comment by OzGav** [comment]
In MA 2.4.0b23 and later mapping of player controls is available.

**Reply by pantherale0** [reply]
Another fantastic update, thank you :)

**Reply by Paul-Vdp** [reply]
Re the Player Controls: 'It allows you for example to redirect the power on/off commands to a different appliance'.
Maybe I'm a bit thick, but I expected to be able to then set the name/entity of this 'different appliance ? Failing this, I have no idea what this setting brings to the table ...
In this context I also wonder what happened to Marcel's idea of 'Add a setting in MA integration to attach a toggle entity to a media player for power control.' (see above), because that is not what these Player Controls seem to be making possible ...

**Reply by pantherale0** [reply]
The power controls can be something like a switch or media player in HA. So when you "turn on" a player in MA, rather than it being simulated, or actually controlling the on/off state of say a snapcast client, it actually controls the switch entity or media player (such as an amp) instead.

**Reply by marcelveldt** [reply]
Exactly - you can assign whatever HA entity you like so you can make it as flexible as you want, especially with the helper entities in HA.

**Reply by Paul-Vdp** [reply]
I get all that - or maybe not, because I fail to see how and where I can assign this HA entity to the media player, or as Marcel once said 'attach a toggle entity to a media player for power control' ...
Cannot find it in the Player's Advanced Settings, nor in the MA integration :-(

**Reply by pantherale0** [reply]
Within MA, head to the plugin settings for Home Assistant and you can expose the entities in the settings there.

**Reply by Paul-Vdp** [reply]
Ok, now we're talking ! Not over-obvious, but who am I ;-)
Anyway, thx for pointing me in the right direction, @pantherale0

**Reply by OzGav** [reply]
This is covered here https://www.music-assistant.io/ha-plugin/
